omnipaxos/src/sequence_paxos/follower.rs

```diff
@@ -190,50 +137,31 @@ where
         }
     }
 
-    pub(crate) fn handle_acceptdecide(&mut self, acc: AcceptDecide<T>) {
-        if self.check_valid_ballot(acc.n)
+    pub(crate) fn handle_acceptdecide(&mut self, acc_dec: AcceptDecide<T>) {
+        if self.check_valid_ballot(acc_dec.n)
             && self.state == (Role::Follower, Phase::Accept)
-            && self.handle_sequence_num(acc.seq_num, acc.n.pid) == MessageStatus::Expected
+            && self.handle_sequence_num(acc_dec.seq_num, acc_dec.n.pid) == MessageStatus::Expected
         {
-            // handle decide
-            let old_decided_idx = self.get_decided_idx();
-            if acc.decided_idx > old_decided_idx {
+            #[cfg(not(feature = "unicache"))]
+            let entries = acc_dec.entries;
+            #[cfg(feature = "unicache")]
+            let entries = self.internal_storage.decode_entries(acc_dec.entries);
+            let mut new_accepted_idx = self
+                .internal_storage
+                .append_entries_and_get_accepted_idx(entries)
+                .expect(WRITE_ERROR_MSG);
+            if acc_dec.decided_idx > self.internal_storage.get_decided_idx() {
+                if acc_dec.decided_idx > self.internal_storage.get_accepted_idx() {
+                    new_accepted_idx =
+                        Some(self.internal_storage.flush_batch().expect(WRITE_ERROR_MSG));
+                }
                 self.internal_storage
-                    .set_decided_idx(acc.decided_idx)
-                    .expect("storage error while trying to write decided index");
+                    .set_decided_idx(acc_dec.decided_idx)
```

Comment:
I think here we should check that the `new_accepted_idx` really was >= the decided index before setting the decided index.

Reply:
Do you thinks its preferable to, instead of flushing the batch, just set decided_idx to `max(accdec.dec_idx, self.internal_storage.get_accepted_idx())`?

Reply:
I prefer to flush. If it is decided then we might as well flush and make it readable for the users.

Reply:
I've ended up doing this:
```rust
    pub(crate) fn handle_acceptdecide(&mut self, acc_dec: AcceptDecide<T>) {
        if self.check_valid_ballot(acc_dec.n)
            && self.state == (Role::Follower, Phase::Accept)
            && self.handle_sequence_num(acc_dec.seq_num, acc_dec.n.pid) == MessageStatus::Expected
        {
            #[cfg(not(feature = "unicache"))]
            let entries = acc_dec.entries;
            #[cfg(feature = "unicache")]
            let entries = self.internal_storage.decode_entries(acc_dec.entries);
            let mut new_accepted_idx = self
                .internal_storage
                .append_entries_and_get_accepted_idx(entries)
                .expect(WRITE_ERROR_MSG);
            let mut new_decided_idx = acc_dec.decided_idx;
            if new_decided_idx > self.internal_storage.get_decided_idx() {
                if new_decided_idx > self.internal_storage.get_accepted_idx() {
                    new_accepted_idx =
                        Some(self.internal_storage.flush_batch().expect(WRITE_ERROR_MSG));
                    new_decided_idx = new_decided_idx.min(new_accepted_idx.unwrap());
                }
                self.internal_storage
                    .set_decided_idx(new_decided_idx)
                    .expect(WRITE_ERROR_MSG);
            }
            if let Some(idx) = new_accepted_idx {
                self.handle_flushed_accepted(acc_dec.n, idx);
            }
        }
    }
```
It feels like we complicating the code for a scenario that should never happen. The leader can never decide entries it hasn't seen itself. So when the follower flushes here it must be the case that its safe to set the `decided_idx` to what the leader told us.

Reply:
Looks good